src/main.rs

```diff
@@ -40,22 +40,41 @@ fn main() {
     let mut action_stack = ActionStack::new();
 
     loop {
-        let cur_action = action_stack.current();
-        let user_selection = UserPrompt::from_str(&user_data.get_user_result(cur_action)).unwrap();
+        let cur_action = action_stack.current().clone();
```

Comment:
it shouldn't be necessary to always `.clone()` that, try if it works without it. NLL's should allow that.